Kconfig.zephyr

```diff
@@ -316,14 +316,29 @@ config CODING_GUIDELINE_CHECK
 	  Use available compiler flags to check coding guideline rules during
 	  the build.
 
-config NATIVE_APPLICATION
-	bool "Build as a native host application"
+config NATIVE_BUILD
+	bool
 	select FULL_LIBC_SUPPORTED
 	select FULL_LIBCPP_SUPPORTED if CPP
+	help
+	  Zephyr will be built targeting the host system for debug and
+	  development purposes.
+
+config NATIVE_APPLICATION
+	bool "Build as a native host application"
+	default y if ARCH_POSIX && !NATIVE_LIBRARY
+	select NATIVE_BUILD
```

Comment:
has `NATIVE_APPLICATION` ever been intended for anything else than the `ARCH_POSIX` ?
(I starting to wonder why the old `NATIVE_APPLICATION` setting has a prompt as ARCH_POSIX has a `select`)

If `NATIVE_APPLICATION` is solely intended for `ARCH_POSIX`, then why not turn the `NATIVE_APPLICATION` and `NATIVE_LIBRARY` into a choice, as they are mutual exclusive afaiu.

Something like:
```
choice NATIVE_CHOICE
	default NATIVE_APPLICATION
	depends on NATIVE_BUILD

config NATIVE_APPLICATION
	bool "Build as a native host application"
	help
	  ...

config NATIVE_LIBRARY
	bool "Build as a prelinked library for the native host target"
	help
	  ...

endchoice
```
```

Reply:
Thanks for taking a look at this PR @tejlmand 
I did consider it being a choice. 
NATIVE_APPLICATION is also selected by unit tests. 
APP and LIBRARY are mutually exclusive, but converting them into a choice, makes them not `select`able, and therefore we cannot select them from the boards. We would only be able to default them, which would carry the risk of a user thinking they can change it.
Therefore I leaned to the current solution.

Reply:
Thanks for your reasoning.
I'm not fully convinced, but that is partly due to the fact that I actually believe the current `NATIVE_APPLICATION` is somewhat flawed.

So good to share thoughts on the choices being made.

> NATIVE_APPLICATION is also selected by unit tests. 

Correct, but just like the arch posix, that could be turned into a select NATIVE_BUILD instead of select NATIVE_APPLICATION.

> APP and LIBRARY are anyhow mutually exclusive, but converting them into a choice, makes them not `select`able, and therefore I cannot select them from the boards

Correct, but they can be specified in `<board>_defconfig`.

Question is, having a prompt on `NATIVE_APPLICATION` (and `NATIVE_LIBRARY`) can give users the false impression that they are able to just enable this on any board / sample they like.
Which, afaik is not the case.

Unsure why `NATIVE_APPLICATION` was given a prompt in the first place, but maybe the author of this PR has some insight https://github.com/zephyrproject-rtos/zephyr/pull/4174 (commit 95cd063280b848f9399272ca7a4c7db8c256977c) 😉 .

Based on what you write and how I read the usage of `NATIVE_APPLICATION` (and `LIBRARY`) in both Kconfig and CMake code, then I believe this should be the choice of the arch, soc, or board, and not something for a user to decide.
Which, as you point out, can be achieved using `select` and that allows us to remove the prompt.

If we do that, we can achieve all we want:
- mutual exclusive
- Native APP is default ( keeping current default behavior)
- Ability to use `select`, which gives board maintainers the ability to use the setting.
- End user cannot mess up the config.

Example:
```
config NATIVE_BUILD
	bool
	.... as now ....

config NATIVE_APPLICATION
	bool
	default y if ARCH_POSIX
	depends on !NATIVE_LIBRARY
	select NATIVE_BUILD
	help
	  Build as a native application that can run on the host and using
	  resources and libraries provided by the host.

config NATIVE_LIBRARY
	bool
	select NATIVE_BUILD
	help
	  Build as a prelinked library for the native host target.
	  This library can later be built into an executable for the host.
```